docs/source/notebooks/probabilistic_matrix_factorization.ipynb

```diff
@@ -0,0 +1,1489 @@
+{
```

Comment:
<p>I would replace this with a function:</p><p> </p><p><code></code><code></p><p>def make_pmf_model(train, dim, alpha, std):</p><p>&nbsp;&nbsp;&nbsp;&nbsp;bounds = (1, 5)</p><p>&nbsp;&nbsp;&nbsp;&nbsp;data = train.copy()</p><p>&nbsp;&nbsp;&nbsp;&nbsp;n, m = data.shape</p><p> </p><p>&nbsp;&nbsp;&nbsp;&nbsp;# Perform mean value imputation</p><p>&nbsp;&nbsp;&nbsp;&nbsp;nan_mask = np.isnan(data)</p><p>&nbsp;&nbsp;&nbsp;&nbsp;data[nan_mask] = data[~nan_mask].mean()</p><p> </p><p>&nbsp;&nbsp;&nbsp;&nbsp;# Low precision reflects uncertainty; prevents overfitting.</p><p>&nbsp;&nbsp;&nbsp;&nbsp;# Set to the mean variance across users and items.</p><p>&nbsp;&nbsp;&nbsp;&nbsp;alpha_u = 1 / data.var(axis=1).mean()</p><p>&nbsp;&nbsp;&nbsp;&nbsp;alpha_v = 1 / data.var(axis=0).mean()</p><p> </p><p>&nbsp;&nbsp;&nbsp;&nbsp;# Specify the model.</p><p>&nbsp;&nbsp;&nbsp;&nbsp;logging.info('building the PMF model')</p><p>&nbsp;&nbsp;&nbsp;&nbsp;with pm.Model() as pmf:</p><p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;U = pm.MvNormal(</p><p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;'U', mu=0, tau=alpha_u * np.eye(dim),</p><p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;shape=(n, dim), testval=np.random.randn(n, dim) * std)</p><p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;V = pm.MvNormal(</p><p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;'V', mu=0, tau=alpha_v * np.eye(dim),</p><p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;shape=(m, dim), testval=np.random.randn(m, dim) * std)</p><p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;R = pm.Normal(</p><p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;'R', mu=(U @ V.T)[~nan_mask], tau=alpha, observed=data[~nan_mask])</p><p>    return pmf</p><p></code><code></code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3709/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>